plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitManager.java

```diff
@@ -108,6 +112,36 @@ public ConnectorSplitSource getSplits(
         return new ClassLoaderSafeConnectorSplitSource(splitSource, IcebergSplitManager.class.getClassLoader());
     }
 
+    private Scan<?, FileScanTask, CombinedScanTask> getScan(IcebergMetadata icebergMetadata, Table icebergTable, IcebergTableHandle table, ExecutorService executor)
+    {
+        Long fromSnapshot = icebergMetadata.getIncrementalRefreshFromSnapshot(table.getSchemaTableName()).orElse(null);
+        if (fromSnapshot != null) {
+            // check if fromSnapshot is still part of the table's snapshot history
+            boolean containsFromSnapshot = SnapshotUtil.isAncestorOf(icebergTable, fromSnapshot);
+            if (containsFromSnapshot) {
+                boolean containsDelete = false;
+                for (Snapshot snapshot : SnapshotUtil.ancestorsBetween(icebergTable, icebergTable.currentSnapshot().snapshotId(), fromSnapshot)) {
+                    if (snapshot.operation().equals(DataOperations.OVERWRITE) || snapshot.operation().equals(DataOperations.DELETE)) {
+                        containsDelete = true;
+                        break;
+                    }
+                }
+                if (!containsDelete) {
+                    return icebergTable.newIncrementalAppendScan().fromSnapshotExclusive(fromSnapshot).planWith(executor);
+                }
+                else {
+                    // the snapshot range contains deletes, so we cannot do an incremental refresh. Fall back to full refresh.
+                    icebergMetadata.disableIncrementalRefresh();
+                }
+            }
+            else {
+                // fromSnapshot is missing (could be due to snapshot expiration or rollback). Fall back to full refresh.
+                icebergMetadata.disableIncrementalRefresh();
```

Comment:
How does the engine/connector knows to replace storage table rather than use append mode?